### PR TITLE
Entity content

### PR DESCRIPTION
**Pull request by huwshimi**
Display content for the bundles and charms.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- view a charm
- check that you get a summary, description and config details
- view a bundle
- check that you get a diagram and description

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3278/
Test PASSed.

**Comment by urosj** [comment]
QA :+1:

**Comment by hatched** [comment]
:+1: QA :ok: Looks awesome!

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

File: jujugui/static/gui/src/app/components/entity-content-diagram/entity-content-diagram.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('entity-content-diagram', function() {
+
+  juju.components.EntityContentDiagram = React.createClass({
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-content-diagram/entity-content-diagram.js]: Is a separate component really needed for this?

File: jujugui/static/gui/src/app/components/entity-content-readme/entity-content-readme.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+      Get the content for the readme.
+
+      @method _getReadme
+      @param {Object} entityModel The entity model.
+      @return {Object} The readme content.
+    */
+    _getReadme: function(entityModel) {
+      var entityModel = this.props.entityModel;
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-content-readme/entity-content-readme.js]: No need to pass in `entityModel` as a function param if we just overwrite it from `this.props`.

File: jujugui/static/gui/src/app/components/entity-content-readme/entity-content-readme.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      Get the filename for the readme.
+
+      @method _getReadmeFile
+      @param {Object} entityModel The entity model.
+      @return {Object} The readme filename.
+    */
+    _getReadmeFile: function(entityModel) {
+      var files = entityModel.get('files');
+      var match;
+      files.some(function(file) {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-content-readme/entity-content-readme.js]: This seems like an odd use of `some` to me. We're not really interested in the boolean returned by `some`, but rather it seems to be used in place of a vanilla for loop and a break statement.

File: jujugui/static/gui/src/app/components/entity-content-readme/test-entity-content-readme.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  beforeEach(function() {
+    mockEntity = makeEntity();
+  });
+
+  afterEach(function() {
+    mockEntity = undefined;
+  });
+
+  function makeEntity(files=[]) {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-content-readme/test-entity-content-readme.js]: This `makeEntity` function is showing up in a lot of test files. Seems like it's a prime candidate to refactor into a shared location.

File: jujugui/static/gui/src/app/components/entity-content/entity-content.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    _generateOptionsList: function(entityModel) {
+      if (entityModel.get('entityType') === 'bundle') {
+        return;
+      }
+      var options = entityModel.get('options');
+      var optionsList = [];
+      Object.keys(options).forEach(function(name) {
+        var option = options[name];
+        optionsList.push(
+          <juju.components.EntityContentConfigOption
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-content/entity-content.js]: Why not just pass `EntityContentConfigOption` the entire `option` object?

File: jujugui/static/gui/src/app/assets/css/_mixins.scss [hunk 6]
```diff
@@ -237,18 +237,15 @@
...
 
-    .header {
-        $shadow-size: 3px;
-        @include create-box-shadow(0px 1px $shadow-size 0px rgba(0, 0, 0, 0.35));
-        margin-bottom: $shadow-size;
-        display: block;
-        padding-top: 30px;
-        padding-bottom: 20px;
-    }
+%fullscreen-panel {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/assets/css/_mixins.scss]: Nicely done.

File: jujugui/static/gui/src/app/components/charmbrowser/charmbrowser.js [hunk 7]
```diff
@@ -86,7 +86,10 @@ YUI.add('charmbrowser-component', function() {
                 importBundleYAML={this.props.importBundleYAML}
                 getBundleYAML={this.props.getBundleYAML}
                 getEntity={this.props.getEntity}
+                getDiagramURL={this.props.getDiagramURL}
                 deployService={this.props.deployService}
+                getFile={this.props.getFile}
+                renderMarkdown={this.props.renderMarkdown}
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/charmbrowser/charmbrowser.js]: I think we could simplify the number of functions being passed in by moving some (particularly these) into the entity model; that is, the entity model would invoke the underlying functions in `charmstore-api.js`, saving us the hassle of passing them down to every component.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/1119

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by kadams54** [comment]
All React components should have `propTypes` defined for any props they need/use.

**Comment by kadams54** [comment]
Created a new card for a follow-up PR that addresses the comments above.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**